proxy/src/coco/peer.rs

```diff
@@ -32,316 +39,401 @@ pub struct UserRevisions {
     }
 
-    let meta: Result<project::Project<entity::Draft>, error::Error> = {
-        // Create the project meta
-        let mut meta = project::Project::<entity::Draft>::create(name.to_string(), owner.urn())?
-            .to_builder()
-            .set_description(description.to_string())
-            .set_default_branch(default_branch.to_string())
-            .add_key(key.public())
-            .add_certifier(owner.urn())
-            .build()?;
-        meta.sign_owned(&key)?;
-        let urn = meta.urn();
-
-        let storage = peer.storage().reopen()?;
-
-        if storage.has_urn(&urn)? {
-            return Err(error::Error::EntityExists(urn));
-        } else {
-            let repo = storage.create_repo(&meta)?;
-            repo.set_rad_self(librad::git::storage::RadSelfSpec::Urn(owner.urn()))?;
+    /// Get the project found at `urn`.
+    ///
+    /// # Errors
+    ///
+    ///   * Resolving the project fails.
+    pub fn get_project(
+        &self,
+        urn: &RadUrn,
+    ) -> Result<project::Project<entity::Draft>, error::Error> {
+        let api = self.peer_api.lock().expect("unable to acquire lock");
+        let storage = api.storage().reopen()?;
+
+        Ok(storage.metadata(urn)?)
+    }
+
+    /// Get the user found at `urn`.
+    ///
+    /// # Errors
+    ///
+    ///   * Resolving the user fails.
+    ///   * Could not successfully acquire a lock to the API.
+    pub fn get_user(&self, urn: &RadUrn) -> Result<user::User<entity::Draft>, error::Error> {
+        let api = self.peer_api.lock().expect("unable to acquire lock");
+        let storage = api.storage().reopen()?;
+
+        Ok(storage.metadata(urn)?)
+    }
+
+    /// Get a repo browser for a project.
+    ///
+    /// # Errors
+    ///
+    /// The function will result in an error if the mutex guard was poisoned. See
+    /// [`std::sync::Mutex::lock`] for further details.
+    pub fn with_browser<F, T>(&self, urn: &RadUrn, callback: F) -> Result<T, error::Error>
+    where
+        F: Send + FnOnce(&mut git::Browser) -> Result<T, error::Error>,
+    {
+        let git_dir = self.monorepo();
+
+        let project = self.get_project(urn)?;
+        let default_branch = git::Branch::local(project.default_branch());
+        let repo = git::Repository::new(git_dir)?;
+        let namespace = git::Namespace::try_from(project.urn().id.to_string().as_str())?;
+        let mut browser = git::Browser::new_with_namespace(&repo, &namespace, default_branch)?;
+
+        callback(&mut browser)
+    }
+
+    /// Initialize a [`project::Project`] that is owned by the `owner`.
+    /// This kicks off the history of the project, tracked by `librad`'s mono-repo.
+    ///
+    /// # Errors
+    ///
+    /// Will error if:
+    ///     * The signing of the project metadata fails.
+    ///     * The interaction with `librad` [`librad::git::storage::Storage`] fails.
+    #[allow(clippy::needless_pass_by_value)] // We don't want to keep `SecretKey` in memory.
```

Comment:
Ideally, we should actually be using the `Signer` trait (see https://github.com/radicle-dev/radicle-link/issues/216). This will delegate the fetch of the secret key from the store and `link` will do it's thing.

So this owned key shenanigans was probably overly cautious in retrospect.

Reply:
#704